pg_analytics/src/hooks/createfunction.rs

```diff
@@ -0,0 +1,155 @@
+use async_std::task;
+use pgrx::*;
+use std::sync::Arc;
+use std::ptr;
+use std::ffi::CStr;
+
+use crate::datafusion::context::DatafusionContext;
+use crate::errors::ParadeError;
+use crate::hooks::handler::DeltaHandler;
+
+use deltalake::datafusion::logical_expr::{ColumnarValue, Volatility, create_udf};
+use deltalake::datafusion::arrow::datatypes::DataType;
+use deltalake::datafusion::common::arrow::array::{ArrayRef, Float64Array};
+use deltalake::datafusion::common::DataFusionError;
+use deltalake::datafusion::execution::FunctionRegistry;
+use deltalake::datafusion::common::ScalarValue;
+use crate::datafusion::datatype::DatafusionTypeTranslator;
+use crate::datafusion::datatype::PostgresTypeTranslator;
+use pgrx::pg_sys::FunctionParameter;
+use pgrx::pg_sys::ListCell;
+use pgrx::pg_sys::NodeTag;
+use pgrx::pg_sys::LookupTypeNameOid;
+use pgrx::pg_sys::NameListToString;
+use pgrx::pg_sys::LookupFuncName;
+use pgrx::pg_sys::String;
+use pgrx::pg_sys::stringToQualifiedNameList;
+use std::ffi::CString;
+use pgrx::pg_sys::Oid;
+use pgrx::pg_sys::OidFunctionCall2Coll;
+use crate::datafusion::datatype::DatafusionMapProducer;
+use pgrx::pg_sys::fmgr_info;
+use std::cmp::max;
+use pgrx::pg_sys::get_func_rettype;
+use pgrx::pg_sys::get_func_signature;
+use crate::errors::NotFound;
+
+pub unsafe fn createfunction(funcname: &str) -> Result<(), ParadeError> {
+    let mut arg_types: *mut pg_sys::Oid = std::ptr::null_mut();
+    let mut nargs: i32 = 0;
+
+    let cstr = CString::new(funcname.clone()).unwrap();
+    let funcname_list = unsafe { stringToQualifiedNameList(cstr.as_ptr(), std::ptr::null_mut()) };
+
+    // TODO: Unless we can guarantee the exact matches for arg Oids, we'll have to assume that only one function
+    // with this name exists. Non-lossy type conversion will be necessary.
+    let func_oid = LookupFuncName(funcname_list, -1, std::ptr::null_mut(), true);
+
+    let ret_oid = get_func_signature(func_oid, &mut arg_types, &mut nargs);
+    if ret_oid == Oid::INVALID {
+        return Err(ParadeError::NotFound(NotFound::Function(funcname.to_string())));
+    }
+
+    // Create vector of input types:
+    let mut input_types = vec![];
+    input_types.push(DataType::Utf8);
+    for param_index in 0..nargs {
+        let arg_oid = arg_types.add(param_index as usize) as *mut Oid;
+        // TODO: typmod?
+        input_types.push(DatafusionTypeTranslator::from_sql_data_type(PgOid::from(*arg_oid).to_sql_data_type(-1)?)?);
+    }
+
+    // TODO: typmod?
+    let return_type = DatafusionTypeTranslator::from_sql_data_type(PgOid::from(ret_oid).to_sql_data_type(-1)?)?;
+
+    let udf_rust = Arc::new(|args: &[ColumnarValue]| -> Result<ColumnarValue, DataFusionError> {
+        let num_args = args.len();
+        let mut num_rows = 1;
+        let mut arg_oids = pg_sys::palloc0(std::mem::size_of::<pg_sys::Oid>() * num_args) as *mut pg_sys::Oid;
```

Comment:
Not 100% about this but I believe `palloc`s should happen in their own Postgres MemoryCtx. Also, is `arg_oids` being used anywhere?

Reply:
o yeah `arg_oids` is not used right now, but when we want to support function overloading, we'll have to use the oids to get the right signature.